services/visual-identity/utils.go

```diff
@@ -0,0 +1,89 @@
+package visualIdentity
+
+import (
+	"errors"
+	"math/big"
+)
+
+func ToBigBase(value *big.Int, base uint64) (res [](uint64)) {
+	toBigBaseImpl(value, base, &res)
+	return res
+}
+
+func toBigBaseImpl(value *big.Int, base uint64, res *[](uint64)) {
+	bigBase := new(big.Int).SetUint64(base)
+	quotient := new(big.Int).Div(value, bigBase)
+	if quotient.Cmp(new(big.Int).SetUint64(0)) != 0 {
+		toBigBaseImpl(quotient, base, res)
+	}
+
+	*res = append(*res, new(big.Int).Mod(value, bigBase).Uint64())
+}
+
+func ToEmojiHash(value *big.Int, hashLen int, alphabet *[]string) (hash []string, err error) {
+	valueBitLen := value.BitLen()
+	alphabetLen := new(big.Int).SetInt64(int64(len(*alphabet)))
+
+	indexes := ToBigBase(value, alphabetLen.Uint64())
+	if hashLen == 0 {
+		hashLen = len(indexes)
+	} else if hashLen > len(indexes) {
+		prependLen := hashLen - len(indexes)
+		for i := 0; i < prependLen; i++ {
+			indexes = append([](uint64){0}, indexes...)
+		}
+	}
+
+	// alphabetLen^hashLen
+	possibleCombinations := new(big.Int).Exp(alphabetLen, new(big.Int).SetInt64(int64(hashLen)), nil)
+
+	// 2^valueBitLen
+	requiredCombinations := new(big.Int).Exp(new(big.Int).SetInt64(2), new(big.Int).SetInt64(int64(valueBitLen)), nil)
+
+	if possibleCombinations.Cmp(requiredCombinations) == -1 {
+		return nil, errors.New("alphabet or hash length is too short to encode given value")
+	}
+
+	for _, v := range indexes {
+		hash = append(hash, (*alphabet)[v])
+	}
+
+	return hash, nil
+}
+
+// compressedPubKey = |1.5 bytes chars cutoff|20 bytes emoji hash|10 bytes color hash|1.5 bytes chars cutoff|
+func Slices(compressedPubkey *[]byte) (res [4][]byte, err error) {
```

Comment:
You can avoid having to dereference the compressedPubKey with `*` if you use `compressedPubkey []byte`. It will pass by copy a small struct, but the underlying byte array is a pointer, so unless you're modifying the compressedPubkey, it should be fine.